File: 032_MaximumTwinSumOfALinkedList.py
"""
Topic: Linked List / Two Pointers
Maximum Twin Sum of a Linked List: reverse second half and compute pair sums
Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def pair_sum(head: Optional[ListNode]) -> int:
    # find middle
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    # reverse second half
    prev = None
    cur = slow
    while cur:
        nxt = cur.next
        cur.next = prev
        prev = cur
        cur = nxt
    # compute max twin sum
    res = 0
    p1, p2 = head, prev
    while p2:
        res = max(res, p1.val + p2.val)
        p1 = p1.next; p2 = p2.next
    return res

def make_list(vals):
    dummy = ListNode(0); cur = dummy
    for v in vals: cur.next = ListNode(v); cur = cur.next
    return dummy.next

if __name__ == '__main__':
    h = make_list([5,4,2,1])
    print(pair_sum(h))  # 6 (5+1 or 4+2 => max 6)
    h2 = make_list([1,100000])
    print(pair_sum(h2))  # 100001

File: 032_MaximumTwinSumOfALinkedList.cpp
// Topic: Linked List / Two pointers
// Maximum Twin Sum of a Linked List: Pair first with last, second with second-last etc and return
// max sum. Time Complexity: O(n) Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;
struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* reverseList(ListNode* head)
{
    ListNode* prev = nullptr;
    while (head)
    {
        ListNode* nxt = head->next;
        head->next = prev;
        prev = head;
        head = nxt;
    }
    return prev;
}

int pairSum(ListNode* head)
{
    // find middle
    ListNode *slow = head, *fast = head;
    while (fast && fast->next)
    {
        slow = slow->next;
        fast = fast->next->next;
    }
    ListNode* second = reverseList(slow);
    int best = 0;
    ListNode* p = head;
    ListNode* q = second;
    while (q)
    {
        best = max(best, p->val + q->val);
        p = p->next;
        q = q->next;
    }
    // optional: restore second half (skipped)
    return best;
}

int main()
{
    ListNode* a = new ListNode(5);
    a->next = new ListNode(4);
    a->next->next = new ListNode(2);
    a->next->next->next = new ListNode(1);
    cout << pairSum(a) << "\n";  // 6 (5+1 or 4+2 -> max 6)
    ListNode* b = new ListNode(1);
    b->next = new ListNode(100000);
    b->next->next = new ListNode(2);
    b->next->next->next = new ListNode(3);
    cout << pairSum(b) << "\n";
    return 0;
}
